frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/StatusManager.java

```diff
@@ -123,8 +136,34 @@ private void pullStatus() throws Exception {
             frameworkName));
       }
     }
+  }
+
+  private Map<String, AggregatedFrameworkStatus> getReusableAggregatedFrameworkStatuses() throws Exception {
+    Map<String, AggregatedFrameworkStatus> reusableAggFrameworkStatuses = new HashMap<>();
+
+    if (aggFrameworkStatuses == null) {
+      return reusableAggFrameworkStatuses;
+    }
+
+    List<FrameworkRequest> allFrameworkRequests = webServer.getAllFrameworkRequests();
+    for (FrameworkRequest frameworkRequest : allFrameworkRequests) {
+      String frameworkName = frameworkRequest.getFrameworkName();
+      Integer frameworkVersion = frameworkRequest.getFrameworkDescriptor().getVersion();
+
+      // For a specific requested FrameworkName and FrameworkVersion, its
+      // AggregatedFrameworkStatus will be immutable after it is transitioned to FINAL_STATES,
+      // so it can be reused directly in next round pullStatus instead of pull from ZK.
+      if (aggFrameworkStatuses.containsKey(frameworkName)) {
+        AggregatedFrameworkStatus aggFrameworkStatus = aggFrameworkStatuses.get(frameworkName);
+        FrameworkStatus frameworkStatus = aggFrameworkStatus.getFrameworkStatus();
+        if (frameworkStatus.getFrameworkVersion().equals(frameworkVersion) &&
+            FrameworkStateDefinition.FINAL_STATES.contains(frameworkStatus.getFrameworkState())) {
```

Comment:
>FINAL_STATES [](start = 37, length = 12)

just to clarify, in your current implementation, the reusable states are all immutable, right?

Reply:
yes, reusable is subset if immutable

---
In reply to: [184586437](https://github.com/Microsoft/pai/pull/534#discussion_r184586437) [](ancestors = 184586437)